Feature/validation system ui trng 1098

SimonTheSourcerer opened the pull request:
Overall architecture changes:
* Switched to EntityData instead of Entities in validation.
* ValidationHandler is a Singleton, LastReport will not be cleared, instead continuously updated with changed parts of the course
* ValidationHandlers validation will be triggered on multiple occasions to reflect the current state of the course.
* TABS got moved to editor, they are not stored in the StepData anymore

Other branches required:
https://github.com/Innoactive/Basic-Interaction-Component/tree/feature/validation-ui-trng-1098
https://github.com/Innoactive/Basic-Conditions-And-Behaviors/tree/feature/validation-trng-374
https://github.com/Innoactive/TextToSpeech-Component/tree/feature/validation-trng-374
https://github.com/Innoactive/Creator-Base-Template/pull/102

SimonTheSourcerer commented on Editor/CourseValidation/ValidationHandler.cs:
```diff
@@ -26,42 +30,55 @@ public ValidationHandler()
         /// <summary>
         /// Validates the given object.
         /// </summary>
         /// <param name="entity">Object, which will be validated.</param>
```
```suggestion
        /// <param name="data">Data object, which will be validated.</param>
```

SimonTheSourcerer commented on Editor/CourseValidation/ValidationHandler.cs:
```diff
@@ -26,42 +30,55 @@ public ValidationHandler()
         /// Validates the given object.
         /// </summary>
         /// <param name="entity">Object, which will be validated.</param>
         /// <param name="course">Course where given <paramref name="entity"/> belongs.</param>
```
```suggestion
        /// <param name="course">Course where given <paramref name="data"/> belongs.</param>
```

SimonTheSourcerer commented on Editor/CourseValidation/ValidationHandler.cs:
```diff
@@ -26,42 +30,55 @@ public ValidationHandler()
         /// <param name="entity">Object, which will be validated.</param>
         /// <param name="course">Course where given <paramref name="entity"/> belongs.</param>
         /// <param name="context">Context of the validation.</param>
         /// <returns>List of reports regarding invalid objects related to the <paramref name="entity"/>.</returns>
```
```suggestion
        /// <returns>List of reports regarding invalid objects related to the <paramref name="data"/>.</returns>
```

SimonTheSourcerer commented on Editor/CourseValidation/ValidationReport.cs:
```diff
@@ -29,17 +30,44 @@ public ValidationReport(List<EditorReportEntry> entries, long generationTime)
             }).ToList();
         }
+
+        public void Update(List<EditorReportEntry> entries, IContext context, long generationTime)
```
xml doc

SimonTheSourcerer commented on Editor/CourseValidation/ValidationUtils.cs:
```diff
@@ -0,0 +1,68 @@
+{
+    internal class ValidationUtils
+    {
+        public static string CreateStepTooltip(List<EditorReportEntry> entries, ChapterContext context)
```
xml doc

SimonTheSourcerer commented on Editor/UI/Graphics/Renderers/StepNodeRenderer.cs:
```diff
@@ -63,6 +66,16 @@ public override void Draw()
+            {
+                GUIContent content = new GUIContent("", null, ValidationUtils.CreateStepTooltip(errors, new ChapterContext(Owner.ActiveChapter.Data, null)));
+                Rect rect = new Rect(Owner.BoundingBox.x + Owner.BoundingBox.width * 0.70f, Owner.BoundingBox.y - 8, 16, 16);
+                GUI.Label(rect, content);
```
probably add a comment why

SimonTheSourcerer commented on Editor/UI/Windows/StepWindow.cs:
```diff
@@ -97,7 +107,7 @@ internal void ResetStepView()
 
             Step currentStep = step as Step;
-            currentStep.Data.Tabs.Selected = default;
+            //currentStep.Data.Tabs.Selected = default;
```
Remove dis

Gusinuhe commented on Editor/CourseValidation/Context/BehaviorContext.cs:
```diff
@@ -3,14 +3,14 @@
         public override bool IsSelectable { get; } = false;
 
-        public BehaviorContext(IBehavior behavior, StepContext parent) : base(behavior, parent) { }
+        public BehaviorContext(IBehaviorData behavior, StepContext parent) : base(behavior, parent) { }
```
Maybe also change the name to `behaviorData`

Gusinuhe replied:
The same for the others that extend from `EntityContext<T>`

Gusinuhe commented on Editor/CourseValidation/Context/ContextExtensions.cs:
```diff
@@ -0,0 +1,44 @@
+﻿namespace Innoactive.CreatorEditor.CourseValidation
+{
+    internal static class ContextExtensions
```
Couldn't this be in a `BaseContext`class? It is odd to have an extension class for our own internal stuff.

SimonTheSourcerer replied:
No, as context is mostly stored as interface reference not as generic class.

Gusinuhe commented on Editor/CourseValidation/Context/EntityContext.cs:
```diff
@@ -27,5 +28,23 @@ public EntityContext(T entity, IContext parent)
         /// <inheritdoc />
         public abstract void Select();
+
+        public override bool Equals(object obj)
```
Missing docs

Gusinuhe commented on Editor/CourseValidation/Context/EntityContext.cs:
```diff
@@ -27,5 +28,23 @@ public EntityContext(T entity, IContext parent)
+            return EqualityComparer<T>.Default.Equals(Entity, other.Entity);
+        }
+
+        public override int GetHashCode()
```
Missing docs

Gusinuhe commented on Editor/CourseValidation/Context/MemberInfoContext.cs:
```diff
@@ -12,11 +13,14 @@ public class MemberInfoContext : IContext
         /// <inheritdoc/>
         public IContext Parent { get; }
 
+        public IData ParentData { get; }
```
Missing docs

Gusinuhe commented on Editor/CourseValidation/Context/MemberInfoContext.cs:
```diff
@@ -12,11 +13,14 @@ public class MemberInfoContext : IContext
 
+        public IData ParentData { get; }
+
         internal MemberInfo MemberInfo { get; }
```
Should this be internal? Public? Is it even used outside this class?

SimonTheSourcerer replied:
yes it is

Gusinuhe commented on Editor/CourseValidation/Context/MemberInfoContext.cs:
```diff
@@ -41,5 +45,23 @@ private string GetMemberInfoName()
             return nameAttribute != null ? nameAttribute.Name : MemberInfo.Name;
         }
+
+        public override bool Equals(object obj)
```
Missing docs

Gusinuhe commented on Editor/CourseValidation/Context/MemberInfoContext.cs:
```diff
@@ -41,5 +45,23 @@ private string GetMemberInfoName()
+            return Equals(MemberInfo, other.MemberInfo);
+        }
+
+        public override int GetHashCode()
```
Missing docs

Gusinuhe commented on Editor/TestTools/EditorImguiTester/EmptyTestStrategy.cs:
```diff
@@ -8,6 +8,8 @@ namespace Innoactive.CreatorEditor.TestTools
     /// </summary>
     internal class EmptyTestStrategy : IEditingStrategy
     {
+        public ICourse CurrentCourse { get; }
```
Shouldn't this be set in the `HandleNewCourseWindow`?

SimonTheSourcerer replied:
Reading the xml documentation, I would say it should do nothing.

Gusinuhe commented on Editor/UI/Drawers/StepDrawer.cs:
```diff
@@ -12,6 +14,41 @@ namespace Innoactive.CreatorEditor.UI.Drawers
     [DefaultTrainingDrawer(typeof(Step.EntityData))]
     internal class StepDrawer : ObjectDrawer
     {
+        public override Rect Draw(Rect rect, object currentValue, Action<object> changeValueCallback, GUIContent label)
```
Missing docs

Gusinuhe commented on Editor/UI/Graphics/StepNode.cs:
```diff
@@ -14,13 +15,16 @@ internal class StepNode : EditorNode
 
         public CreateTransitionButton CreateTransitionButton { get; private set; }
 
+        public IChapter ActiveChapter { get; }
```
Missing docs

Gusinuhe commented on Editor/UI/Tabs/DynamicTab.cs:
```diff
@@ -1,15 +1,15 @@
 using UnityEngine;
 
-namespace Innoactive.Creator.Core.Tabs
+namespace Innoactive.CreatorEditor.Tabs
```
😱🤞

Gusinuhe commented on Editor/UI/Tabs/ITabsGroup.cs:
```diff
@@ -11,7 +10,6 @@ internal interface ITabsGroup
         /// <summary>
         /// Index of the currently selected tab.
         /// </summary>
-        [DataMember]
```
Why was it even there?

Gusinuhe commented on Editor/UI/Windows/StepWindow.cs:
```diff
@@ -51,6 +53,14 @@ private void OnInspectorUpdate()
 
+        void OnFocus()
+        {
+            if (RuntimeConfigurator.Exists)
```
It would be nice to be able to somehow disable the validation system.

Gusinuhe commented on Editor/UI/Windows/TrainingMenuView.cs:
```diff
@@ -136,28 +136,6 @@ public void Draw()
                 GUILayout.BeginVertical("box");
                 {
                     EditorColorUtils.ResetBackgroundColor();
-
-                    if (GUI.Button(new Rect(ChapterPaddingTop, ChapterPaddingTop, 120, 28), "Validate Course"))
-                    {
-                        ValidationReport report = validation.Validate(Course, Course);
-                        report.Entries.ForEach(entry =>
-                        {
-                            if (entry.ErrorLevel == ValidationErrorLevel.HINT)
-                            {
-                                Debug.Log(entry.ToString());
-                            }
-                            else if (entry.ErrorLevel == ValidationErrorLevel.WARNING)
-                            {
-                                Debug.LogWarning(entry.ToString());
-                            }
-                            else if (entry.ErrorLevel == ValidationErrorLevel.ERROR || entry.ErrorLevel == ValidationErrorLevel.FATAL)
-                            {
-                                Debug.LogError(entry.ToString());
-                            }
-                        });
-                        Debug.Log($"Validation completed in {report.GenerationTime}ms");
-                    }
-
```
🎉

SimonTheSourcerer commented:
* Added proper check if validation should run, added the possibility to disable validation.
* Removed select completly as it's still not used
* Fixed xml doc

Maxsiom reviewed (approved):
Looks good to me. Tested it thoroughly with every behavior and condition. Seems to catch all error. 
Two things that I stumbled across:
- Chapter warnings do not have any tooltip. Wanted?
- Warning begin with the character "*". Any reason for that?

SimonTheSourcerer commented:
Can be more than one error so this should indicate it's a "list" 

Chapters dont have a tooltip because these are only the warning that there are errors in this chapter. So it's on purpose right now.

tomwim commented on .Documentation/articles/developer/14-validation.md:
```diff
@@ -0,0 +1,17 @@
+
+For validating Behaviors and Condition Data we implemented the support of attribute validators, which can be used to validate specific fields.
+
+The `StepAttributeValidator` checks for all Behaviors & Condition if there are any Attributes added which implement the interface `IAttributeValidator`. All found validators will be run and the reports given back to the ValidationHandler.
```
```suggestion
The `StepAttributeValidator` checks for all Behaviors & Conditions if there are any Attributes added which implement the interface `IAttributeValidator`. All found validators will be run and the reports given back to the ValidationHandler.
```

tomwim commented on .Documentation/articles/developer/14-validation.md:
```diff
@@ -0,0 +1,17 @@
+
+## Attribute Validation
+
+For validating Behaviors and Condition Data we implemented the support of attribute validators, which can be used to validate specific fields.
```
```suggestion
For validating Behavior and Condition Data we implemented the support of attribute validators, which can be used to validate specific fields.
```

tomwim commented on Editor/CourseValidation/ValidationHandler.cs:
```diff
@@ -29,39 +36,55 @@ public ValidationHandler()
-        /// <returns>List of reports regarding invalid objects related to the <paramref name="entityObject"/>.</returns>
-        protected List<EditorReportEntry> InternalValidate(object entityObject, IContext context)
+        /// <returns>List of reports regarding invalid objects related to the <paramref name="validateableObject"/>.</returns>
+        protected List<EditorReportEntry> InternalValidate(object validateableObject, IContext context)
```
```suggestion
        protected List<EditorReportEntry> InternalValidate(object validatableObject, IContext context)
```
It is actually _validatable_ without the 'e'

tomwim commented on Editor/CourseValidation/ValidationUtils.cs:
```diff
@@ -0,0 +1,74 @@
+namespace Innoactive.CreatorEditor.CourseValidation
+{
+    /// <summary>
+    /// Validation utilities, this most generous xml doc was done so every is super happy.
```
hahahahah...dont think that is acceptable. It is also grammatically wrong :D

tomwim replied:
```suggestion
    /// Validation utilities.
```

tomwim commented on Runtime/CourseValidation/CheckForColliderAttribute.cs:
```diff
@@ -9,7 +11,12 @@ public class CheckForColliderAttribute : CheckForComponentAttribute
+
+        }
+
+        protected override ReportEntry CreateErrorMessage(GameObject gameObject, List<Type> components)
```
inhertidoc

tomwim commented:
Behaviors in BehaviorSequence are not validated.

Probably not part of the validation system but behaviors/conditions with a time that can be set (e.g. delay or move object into collider) should have a minimum time

SimonTheSourcerer commented:
Seems BehaviorSequence needs special attention to work. I will take a look, hopefully easy.

Minimum time should already be implemented.

Luphoxa commented:
Part 1
### Bugs I found:
The Unlocked Objects Tab is broken with these branches while it still works in develop. Objects seem to be unlocked if referenced though.

### Questions I would like to have answered:

1.  What if an invalid localization key is given? The creator would know once the project is started once, right? Is there the possibility to add a warning then?
2. Behavior Sequences have no warnings, will this be covered in a later update or should it be part of this increment? The audio hint probably has no warnings for the same reason, right?
3. Is there a possibility to make error messages more precise, like in this case, where the value is given, but not configured right.
![image](https://user-images.githubusercontent.com/15690875/95863350-5d693a00-0d64-11eb-984e-a1efd93ebaf9.png)
4. The Animation Duration in Scale Object probably could also be optional, while I guess the target scale should be required, what do you think?
![image](https://user-images.githubusercontent.com/15690875/95864101-40813680-0d65-11eb-95d7-d2a47a3346c5.png)
5. What happens if the object has the scrips, but they are disabled? Does this impact the training? Should we check that they are indeed enabled?
6. Move Object into Collider: required seconds should be optional. (If they aren't already)
7. Should we check if people use the same object in a behavior/condition where different objects should be used? Example
![image](https://user-images.githubusercontent.com/15690875/95866008-a53d9080-0d67-11eb-9d04-aabedc61cb89.png)

Luphoxa commented:
Part 2 
### Questions & Remarks
8. Required Seconds Inside of _Object Nearby_ probably could also be optional
9. Can we check if a collider is set to _IsTrigger_ or not with the validation system? Move object into collider and snap zone require those, but there is no warning if set wrong as far as I could see

SimonTheSourcerer commented:
BehaviorSequence not rendering is a missed part of the implementation. I will check out if it is easy to implement.

Missing additional checks are not bugs, they are new features which are not required for this implementation. They still sound like a good idea and should be kept around to implement them later.

And the optional things should probably be optional yes :D

SimonTheSourcerer commented:
- Added BehaviorSequence to validation
- Added missing inerhit doc
- Errors while validating are only shown when debug validation is flagged true
- Extended is allowed to validate

Gusinuhe commented on .Documentation/articles/developer/14-validation.md:
```diff
@@ -0,0 +1,36 @@
+
+**What do have to take into account to get validation properly working?**
+
+- The validation system can be enabled/disabled by switchting the CreatorProjectSettings entry `IsValidationEnabled`.
```
And where do I get or find `CreatorProjectSettings`?

Gusinuhe commented on .Documentation/articles/developer/14-validation.md:
```diff
@@ -0,0 +1,36 @@
+**What do have to take into account to get validation properly working?**
+
+- The validation system can be enabled/disabled by switchting the CreatorProjectSettings entry `IsValidationEnabled`.
+- Validation only properly works with the correct scene opend in the unity editor.
```
I would say that it also works having the incorrect scene/training course opened 🤔

SimonTheSourcerer replied:
And it also shows a lot of reference errors which, in the correct scene, are working as intended.

Gusinuhe commented on Editor/CourseValidation/ValidationReport.cs:
```diff
@@ -29,17 +32,59 @@ public ValidationReport(List<EditorReportEntry> entries, long generationTime)
             }).ToList();
         }
+
+        public List<EditorReportEntry> GetEntriesFor(IBehaviorData data)
```
Missing docs

SimonTheSourcerer replied:
internal class........

Gusinuhe commented on Editor/CourseValidation/ValidationReport.cs:
```diff
@@ -29,17 +32,59 @@ public ValidationReport(List<EditorReportEntry> entries, long generationTime)
+            return Entries.Where(entry => entry.Context.Equals(context)).ToList();
+        }
+
+        public List<EditorReportEntry> GetEntriesFor(IConditionData data)
```
Missing docs

SimonTheSourcerer replied:
internal class........

Gusinuhe commented on Editor/CreatorProjectSettings.cs:
```diff
@@ -13,6 +13,18 @@ public class CreatorProjectSettings : ScriptableObject
     [SerializeField]
     public bool IsFirstTimeStarted = true;
 
+    /// <summary>
+    /// Enables the validation system.
+    /// </summary>
+    [SerializeField]
+    public bool IsValidationEnabled = true;
+
+    /// <summary>
+    /// Enables the validation system.
+    /// </summary>
+    [SerializeField]
+    public bool DebugValidationSystem = false;
```
So, both enable the validation system? 🤔

Gusinuhe commented on Editor/DefaultEditingStrategy.cs:
```diff
@@ -13,7 +15,8 @@ internal class DefaultEditingStrategy : IEditingStrategy
         private StepWindow stepWindow;
-        private ICourse course;
+
+        public ICourse CurrentCourse { get; protected set; }
```
Missing docs

SimonTheSourcerer replied:
it's an internal class

Gusinuhe commented on Editor/UI/Drawers/DrawerLocator.cs:
```diff
@@ -4,6 +4,7 @@
 using System.Reflection;
 using Innoactive.Creator.Core.Attributes;
 using Innoactive.Creator.Core.Utils;
+using UnityEngine;
```
Is this required?

Gusinuhe commented on Runtime/CourseValidation/CheckForColliderAttribute.cs:
```diff
@@ -7,9 +9,16 @@ namespace Innoactive.Creator.Core.Validation
         public CheckForColliderAttribute() : base(typeof(BoxCollider), typeof(SphereCollider), typeof(CapsuleCollider), typeof(MeshCollider), typeof(Collider))
         {
-            message = "Target TrainingSceneObject is missing a collider!";
+
+        }
```
```suggestion
        }
```

Gusinuhe commented on Runtime/CourseValidation/CheckForColliderAttribute.cs:
```diff
@@ -7,9 +9,16 @@ namespace Innoactive.Creator.Core.Validation
     /// </summary>
     public class CheckForColliderAttribute : CheckForComponentAttribute
     {
+        /// <inheritdoc />
```
```suggestion
```

Gusinuhe commented on Runtime/CourseValidation/ReportEntryGenerator.cs:
```diff
@@ -0,0 +1,42 @@
+
+namespace Innoactive.Creator.Core.Validation
+{
+    public static class ReportEntryGenerator
```
Missing docs in the whole class

Gusinuhe reviewed (changes requested):
How can anyone access `.Documentation/articles/developer/14-validation.md`?
